pkg/image/external/external.go

```diff
@@ -0,0 +1,154 @@
+package external
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strings"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/rancher/rancher/pkg/controllers/management/k3sbasedupgrade"
+	"github.com/sirupsen/logrus"
+)
+
+type Source string
+
+const (
+	K3S  Source = "k3s"
+	RKE2 Source = "rke2"
+)
+
+func GetExternalImages(rancherVersion string, externalData map[string]interface{}, source Source, minimumKubernetesVersion *semver.Version) ([]string, error) {
+	if source != K3S && source != RKE2 {
+		return nil, fmt.Errorf("invalid source provided: %s", source)
+	}
+
+	logrus.Infof("generating %s image list...", source)
+	externalImagesMap := make(map[string]bool)
+	releases, _ := externalData["releases"].([]interface{})
+
+	var compatibleReleases []string
+
+	for _, release := range releases {
+		releaseMap, _ := release.(map[string]interface{})
+		version, _ := releaseMap["version"].(string)
+		if version == "" {
+			continue
+		}
+
+		// Skip the release if a minimum Kubernetes version is provided and is not met.
+		if minimumKubernetesVersion != nil {
+			versionSemVer, err := semver.NewVersion(strings.TrimPrefix(version, "v"))
+			if err != nil {
+				continue
+			}
+			if versionSemVer.LessThan(*minimumKubernetesVersion) {
+				continue
+			}
+		}
+
+		if rancherVersion != "dev" {
+			maxVersion, _ := releaseMap["maxChannelServerVersion"].(string)
+			maxVersion = strings.TrimPrefix(maxVersion, "v")
+			if maxVersion == "" {
+				continue
+			}
+			minVersion, _ := releaseMap["minChannelServerVersion"].(string)
+			minVersion = strings.Trim(minVersion, "v")
+			if minVersion == "" {
+				continue
+			}
+
+			versionGTMin, err := k3sbasedupgrade.IsNewerVersion(minVersion, rancherVersion)
+			if err != nil {
+				continue
+			}
+			if rancherVersion != minVersion && !versionGTMin {
+				// Rancher version not equal to or greater than minimum supported rancher version.
+				continue
+			}
+
+			versionLTMax, err := k3sbasedupgrade.IsNewerVersion(rancherVersion, maxVersion)
+			if err != nil {
+				continue
+			}
+			if rancherVersion != maxVersion && !versionLTMax {
+				// Rancher version not equal to or greater than maximum supported rancher version.
+				continue
+			}
+		}
+
+		logrus.Debugf("[%s] adding compatible release: %s", source, version)
+		compatibleReleases = append(compatibleReleases, version)
+	}
+
+	if compatibleReleases == nil || len(compatibleReleases) < 1 {
+		logrus.Infof("skipping image generation since no compatible releases were found for version: %s", rancherVersion)
+		return nil, nil
+	}
```

Comment:
Requested by @superseb for clarity on incompatible runs (like commits to `master` for PRs).